feat(webhooks): Add support for a userid filter

come-nc opened the pull request:
## Summary

This allows to register a userId to filter on along with the webhooks. This webhook will then only be triggered if the given userId is the one in session. This is more efficient than filtering by user in the event filter because the listener is not even registered if the user id does not match.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

come-nc committed:
feat(webhooks): Add support for a userid filter

This allows to register a userId to filter on along with the webhooks.
This webhook will then only be triggered if the given userId is the one
in session. This is more efficient than filtering by user in the event
filter because the listener is not even registered if the user id does
not match.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(webhooks): Fix userIdFiltering for webhooks calls

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
I did not test it in real conditions yet, but code is ready for review.

come-nc committed:
fix(webhooks): Completed API documentation and updated openapi.json

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

bigcat88 commented on apps/webhook_listeners/lib/Db/WebhookListenerMapper.php:
```diff
@@ -181,27 +189,38 @@ private function getAllConfiguredEventsFromDatabase(): array {
 	}
+
+	private function buildCacheKey(?string $userIdFilter = ''): string {
+		return self::EVENTS_CACHE_KEY_PREFIX.'_'.($userIdFilter ?? '');
```
The null coalescing operator (??) is used to provide a default value if the variable is null. 

Since the default value of $userIdFilter is an empty string (''), and an empty string is not null, the null coalescing operator here is unnecessary.

bigcat88 commented on apps/webhook_listeners/lib/Migration/Version1000Date20240527153425.php:
```diff
@@ -53,6 +53,10 @@ public function changeSchema(IOutput $output, Closure $schemaClosure, array $opt
 				'notnull' => false,
 			]);
+			$table->addColumn('user_id_filter', Types::STRING, [
+				'notnull' => true,
```
When using Oracle database the empty value is `null`, afaik..

marcelklehr commented on apps/webhook_listeners/lib/AppInfo/Application.php:
```diff
@@ -40,9 +41,10 @@ private function registerRuleListeners(
 
 		/* Listen to all events with at least one webhook configured */
-		$configuredEvents = $mapper->getAllConfiguredEvents();
+		$configuredEvents = $mapper->getAllConfiguredEvents($userSession->getUser()?->getUID());
```
This potentially calls the method with `null` but the method doesn't accept null

come-nc replied:
The method does accept null

come-nc committed:
fix(webhook_listeners): Allow null for user_id_filter for oracle

Empty string and null are the same thing on oracle databases so we have
 to allow it.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(webhooks_listeners): Fix SQL requests for oracle

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(webhook_listeners): Use string type instead of text for event

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>
